### DISCUSSION TITLE
How to use express.js JWT token?

### DISCUSSION THREAD

**Discussion by somersby10ml** [nextauthjs/next-auth#8150] [Help]
I am using 2 servers.

### express.js (auth server)
```
const accessToken = jwt.sign(payload, SECRET, { expiresIn: '10day' });
res.json({ accessToken }).end();
```

### next.js (+next-auth)
```
export const authOptions: AuthOptions = {
   // session: {
   // maxAge: 60 * 60,
   // },

   providers: [
     CredentialsProvider({
       credentials: {
         username: {},
         password: {},
       },
       authorize: async (credentials) => {
         if (!credentials) {
           return null;
         }

         const res = await fetch('http://localhost:3000/auth/signin', {
           method: 'POST',
           headers: { 'Content-Type': 'application/json' },
           body: JSON.stringify(credentials),
         });
         const data = await res.json();
         const accessToken = data. accessToken as string;
         const decodedToken = jwt. verify(accessToken, NEXTAUTH_SECRET) as User;

         const q: User = {
           ...decodedToken,
           accessToken,
         };
         return q;
       },
     }),
   ],
   callbacks: {
    jwt: async ({ token, user }) => {
      if (user) {
        token.user = {
          id: user.id,
          username: user.username,
        };
        token.accessToken = user.accessToken;
      }
      return token;
    },
    session: async ({ session, token }) => {
      if (token && session.user) {
        session.user = token.user;
        session.accessToken = token.accessToken;
      }
      // console.log(session);
      return session;
    },

  },
};
```

If you run the above source
It looks like the cookie's next-auth.session-token contains the newly issued JWT token from next-auth.

This way, when the express.js token expires
need to be reissued

It looks like the next-auth token may not expire.
So I want to unify the tokens into one.

Is there a way to use a token issued by express.js in next-auth?

**Comment by future-leader1** [comment]
A handy trick here is that NextAuth lets you specify your own signing key for JWTs, which can be pretty useful if you're already creating JWTs from another server. Essentially, you can share the same secret key between your Express.js and NextAuth setups.


Express.js Server: When you create your JWTs in your Express.js server, make sure to use a specific secret key, like this:
javascript
Copy code
const jwt = require('jsonwebtoken');
const SECRET = 'your-secret-key';

const payload = {
  id: 'user-id',
  username: 'user-name'
};

const accessToken = jwt.sign(payload, SECRET, { expiresIn: '10day' });

res.json({ accessToken }).end();
NextAuth Configuration: Now, in your NextAuth setup, ensure you use this same secret key for signing and verifying JWTs. like below 

import NextAuth from 'next-auth'
import Providers from 'next-auth/providers'

const options = {
  providers: [
  ],
  callbacks: {
    jwt: async ({ token, user }) => {
      if (user) {
        token = {
          ...token,
          id: user.id,
          username: user.username,
          accessToken: user.accessToken,
        };
      }
      return token;
    },
    session: async ({ session, token }) => {
      if (token) {
        session.user = {
          id: token.id,
          username: token.username,
        };
        session.accessToken = token.accessToken;
      }
      return session;
    },
  },
  secret: 'your-secret-key', 
  session: {
    jwt: true,
  },
}

export default (req, res) => NextAuth(req, res, options)
 It'll then employ the secret you provide to handle the signing and verifying of JWTs.



Lastly,if you duntno,  remember that you'll need to handle token expiration yourself. When the token expires, it's a good idea to redirect the user to the login page or trigger a silent refresh of the token. NextAuth doesn't handle token refresh automatically.

**Reply by getellez** [reply]
How would the token refresh be handled? Should we use a long-term access token or should we refresh the access token and update the session object every time we refresh it?

**Reply by future-leader1** [reply]
hey man, 

**Long-lived Refresh Tokens:** You use two tokens - a short-lived "access token" for regular use, and a long-lived "refresh token." When the access token expires, the client uses the refresh token to get a new access token without making the user log in again.

**Silent Token Refresh:** You use a single access token, and before it expires, the client gets a new one automatically in the background without bothering the user. It's simpler to set up but might be a bit less secure.

Which one to use depends on how much control and complexity you want. The first one gives more control, the second one is simpler.
